File: TransCoder_tokenized_test_set_functions/python/FIND_MINIMUM_SHIFT_LONGEST_COMMON_PREFIX.py
def KMP ( m, n, str2, str1 ) :
    pos = 0
    Len = 0
    p = [ 0 for i in range ( m + 1 ) ]
    k = 0
    for i in range ( 2, n + 1 ) :
        while ( k > 0 and str1 [ k ] != str1 [ i - 1 ] ) :
            k = p [ k ]
        if ( str1 [ k ] == str1 [ i - 1 ] ) :
            k += 1
        p [ i ] = k
    j = 0
    for i in range ( m ) :
        while ( j > 0 and j < n and str1 [ j ] != str2 [ i ] ) :
            j = p [ j ]
        if ( j < n and str1 [ j ] == str2 [ i ] ) :
            j += 1
        if ( j > Len ) :
            Len = j
            pos = i - j + 1
    print ( "Shift = ", pos )
    print ( "Prefix = ", str1 [ : Len ] )

File: TransCoder_tokenized_test_set_functions/java/FIND_MINIMUM_SHIFT_LONGEST_COMMON_PREFIX.java
static void KMP ( int m, int n, String str2, String str1 ) {
  int pos = 0, len = 0;
  int [ ] p = new int [ m + 1 ];
  int k = 0;
  char [ ] ch1 = str1 . toCharArray ( );
  char [ ] ch2 = str2 . toCharArray ( );
  for ( int i = 2;
  i <= n;
  i ++ ) {
    while ( k > 0 && ch1 [ k ] != ch1 [ i - 1 ] ) k = p [ k ];
    if ( ch1 [ k ] == ch1 [ i - 1 ] ) ++ k;
    p [ i ] = k;
  }
  for ( int j = 0, i = 0;
  i < m;
  i ++ ) {
    while ( j > 0 && j < n && ch1 [ j ] != ch2 [ i ] ) j = p [ j ];
    if ( j < n && ch1 [ j ] == ch2 [ i ] ) j ++;
    if ( j > len ) {
      len = j;
      pos = i - j + 1;
    }
  }
  System . out . println ( "Shift = " + pos );
  System . out . println ( "Prefix = " + str1 . substring ( 0, len ) );
}